src/Microsoft.OData.Edm/ExtensionMethods/EdmTypeSemantics.cs

```diff
@@ -1216,18 +1221,36 @@ public static EdmPrimitiveTypeKind PrimitiveKind(this IEdmTypeReference type)
         /// <param name="potentialBaseType">The potential base type of the type being tested.</param>
         /// <returns>True if and only if the type inherits from the potential base type.</returns>
         public static bool InheritsFrom(this IEdmStructuredType type, IEdmStructuredType potentialBaseType)
+        {            
+            string fullName = type.FullTypeName();
+            HashSet<IEdmStructuredType> baseTypes = GetBaseTypes(type);
+
+            return baseTypes.Contains(potentialBaseType);
+        }
+
+        private static HashSet<IEdmStructuredType> GetBaseTypes(IEdmStructuredType type)
         {
-            do
+            HashSet<IEdmStructuredType> baseTypes;
+
+            if (baseTypeCache.TryGetValue(type, out baseTypes))
+            {
+                return baseTypes;
+            }
+
+            baseTypes = new HashSet<IEdmStructuredType>();
+            baseTypeCache.TryAdd(type, baseTypes);

```

Comment:
We generally add the item to the hashset after it's fully populated. Hashset is not concurrent, so if we are adding items to baseTypes while people are trying to read it that could be a problem.